// @api-platform/admin/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Show = exports.List = exports.resourceFactory = exports.inputFactory = exports.fieldFactory = exports.Edit = exports.Create = exports.AdminBuilder = undefined;

var _hydra = require('./hydra');

Object.keys(_hydra).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _hydra[key];
    }
  });
});

var _AdminBuilder2 = require('./AdminBuilder');

var _AdminBuilder3 = _interopRequireDefault(_AdminBuilder2);

var _Create2 = require('./Create');

var _Create3 = _interopRequireDefault(_Create2);

var _Edit2 = require('./Edit');

var _Edit3 = _interopRequireDefault(_Edit2);

var _fieldFactory2 = require('./fieldFactory');

var _fieldFactory3 = _interopRequireDefault(_fieldFactory2);

var _inputFactory2 = require('./inputFactory');

var _inputFactory3 = _interopRequireDefault(_inputFactory2);

var _resourceFactory2 = require('./resourceFactory');

var _resourceFactory3 = _interopRequireDefault(_resourceFactory2);

var _List2 = require('./List');

var _List3 = _interopRequireDefault(_List2);

var _Show2 = require('./Show');

var _Show3 = _interopRequireDefault(_Show2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.AdminBuilder = _AdminBuilder3.default;
exports.Create = _Create3.default;
exports.Edit = _Edit3.default;
exports.fieldFactory = _fieldFactory3.default;
exports.inputFactory = _inputFactory3.default;
exports.resourceFactory = _resourceFactory3.default;
exports.List = _List3.default;
exports.Show = _Show3.default;
